roles: introduce role for http servers (#197)
tarantool/http: README.md

[socket_ref]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-tcp-server

## Roles

> New roles could be accessed from this project.

Reviewer: ```suggestion
Tarantool 3 roles could be accessed from this project.
```

### `roles.httpd`

> This role can be used only with Tarantool 3. It allows configuring one or more
> HTTP servers. Those servers could be reused by several other roles.

Reviewer: ```suggestion
It allows configuring one or more HTTP servers. Those servers could be reused by several
other roles.
```

```

Server address should be provided either as a URI or as a single port
(in this case, `0.0.0.0` address is used).

User can access every working HTTP server from the configuration by name,
using `require('roles.httpd').get_server(name)` method.
If the `name` argument is `nil`, the default server is returned
(its name should be equal to constant
`require('roles.httpd').DEFAULT_SERVER_NAME`, which is `"default"`).

Let's look at the example of using this role. Consider a new role
`roles/hello_world.lua`:
```lua
local M = { dependencies = { 'roles.httpd' } }
local server = {}

M.validate = function(conf)
    if conf == nil or conf.httpd == nil then
        error("httpd must be set")
    end
    local server = require('roles.httpd').get_server(conf.httpd)
    if server == nil then
        error("the httpd server " .. conf.httpd .. " not found")
    end
end

M.apply = function(conf)
    server = require('roles.httpd').get_server(conf.httpd)

    server:route({
        path = '/hello/world',
        name = 'greeting',
    }, function(tx)
        return tx:render({text = 'Hello, world!'})
    end)
end

M.stop = function()
    server:delete('greeting')
end

> 

Reviewer: Missed dependency:

https://github.com/tarantool/doc/issues/3883

Reviewer: Still no dependency.

[socket_ref]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-tcp-server

## Roles

New roles could be accessed from this project.

### `roles.httpd`

This role can be used only with Tarantool 3. It allows configuring one or more
HTTP servers. Those servers could be reused by several other roles.

Example of the configuration:

```yaml
roles_cfg:
  roles.httpd:
    default:
    - listen: 8081
    additional:
    - listen: '127.0.0.1:8082'
```

Server address should be provided either as a URI or as a single port
(in this case, `0.0.0.0` address is used).

User can access every working HTTP server from the configuration by name,
using `require('roles.httpd').get_server(name)` method.
If the `name` argument is `nil`, the default server is returned
(its name should be equal to constant
`require('roles.httpd').DEFAULT_SERVER_NAME`, which is `"default"`).

Let's look at the example of using this role. Consider a new role
`roles/hello_world.lua`:
```lua
local M = {}

> local server = {}

Reviewer 2: The reason to cache it here is to do a `stop` cleanup?

Author: Yes.

Reviewer: Still no dependency.

M.validate = function()  end

M.apply = function(conf)

>     server = require('roles.httpd').get_server(conf.httpd).httpd

Reviewer 2: > get_server().httpd

Why don't we return `httpd` itself? If I got it right, the `server` consists of `httpd` and `routes`, while `routes` is some internal `role.httpd` thing that is not expected to be used or changed by external roles.

> local M = {}
> local server = {}
> 
> M.validate = function()  end
> 
> M.apply = function(conf)
>     server = require('roles.httpd').get_server(conf.httpd).httpd
> 
>     server:route({
>         path = '/hello/world',
>         name = 'greeting',
>     }, function(tx)
>         return tx:render({text = 'Hello, world!'})
>     end)
> end
> 
> M.stop = function()
>     server:delete('greeting')
> end
> 
> return M

Reviewer: ```suggestion
local M = {}

M.validate = function(conf)
    if conf == nil or conf.httpd == nil then
        error("httpd must be set")
    end
    local server = require('roles.httpd').get_server(conf.httpd).httpd
    if server == nil then
        error("the httpd server " .. conf.httpd .. " not found")
    end
end

M.apply = function(conf)
    local server = require('roles.httpd').get_server(conf.httpd).httpd

    server:route({
        path = '/hello/world',
        name = 'greeting',
    }, function(tx)
        return tx:render({text = 'Hello, world!'})
    end)
end

M.stop = function()
    local server = require('roles.httpd').get_server(conf.httpd).httpd    
    server:delete('greeting')
end

return M
```

Reviewer: It would be nice to add a dependency `roles.httpd` for the role too:
https://www.tarantool.io/en/doc/latest/platform/app/app_roles/#role-dependencies

```

This role accepts a server by name from a config and creates a route to return 
`Hello, world!` to every request by this route.

Then we need to write a simple config to start the Tarantool instance via
`tt`:
```yaml
app:
  file: 'myapp.lua'

groups:
  group001:
    replicasets:
      replicaset001:
        roles: [roles.httpd, roles.hello_world]
        roles_cfg:
          roles.httpd:
            default:
              listen: 8081
            additional:
              listen: '127.0.0.1:8082'
          roles.hello_world:
            httpd: 'additional'
        instances:
          instance001:
            iproto:
              listen:
                - uri: '127.0.0.1:3301'
```

In `myapp.lua` we need to write a script to initialize both roles:
```lua
local httpd_role = require('roles.httpd')
local hello_role = require('roles.hello_world')

httpd_role.apply(config.roles_cfg.httpd)
hello_role.apply(config.roles_cfg.hello_world)
```

> 

Reviewer: It could be removed, just config.yaml + tt start should be enough.

Reviewer 2: Yes, `role.apply` is not designed to be called directly in applications/other roles. (We may do so in unit tests, but just for the sake of unit testing.)

[socket_ref]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-tcp-server

## Roles

New roles could be accessed from this project.

### `http_server` role

This role can be used only with Tarantool 3.
It allows configuring one or more HTTP servers.
Those servers could be reused by several other roles.

Example of the configuration:

```yaml
roles_cfg:
  roles.http_server:
    default:
    - listen: 8081
    additional:
    - listen: '127.0.0.1:8082'
```

Server address should be provided either as a URI or as a single port
(in this case, `0.0.0.0` address is used).

User can access every working HTTP server from the configuration by name,
using `get_server(name)` method.
If the `name` argument is `nil`, the default server is returned
(its name should be equal to constant `DEFAULT_SERVER_NAME`).

For an example of using this role, refer to

> [this integration test](./test/integration/http_server_role_test.lua).

Author: Right now no other project supports this role. And to show an actual example, AFAIU, we either need to show how to support this role from the other role (which is done in `./test/integration/http_server_role_test.lua` via custom mock role), or just write some pulled from thin air example directly in `README`.

For instance, the second way will look like this:
`To show an example on how to use this role with other roles, lets consider role *my_role*, which uses HTTP servers for its own purposes. To configure the connection to each server, the config expects the name of the server and some endpoints to connect. In this example, config file may look like this:`
```yml
roles_cfg:
  roles.http_server:
    default:
    - listen: 8081
    additional:
    - listen: '127.0.0.1:8082'
  roles.my-role:
    - endpoints:
      - path: /metrics
      - path: /metrics/prometheus/
        format: prometheus
    - server: additional
      endpoints:
      - path: /metrics/
```
`Notice that the first connection will use the default server, since no name is provided.`

Or something like that. Will this be ok to put as the doc?

> [this integration test](./test/integration/http_server_role_test.lua).

Reviewer: I would like to see a code of minimal role that output "Hello world" + an example of the configuration for the role:
```yaml
roles_cfg:
  roles.http_server:
    default:
    - listen: 8081
    additional:
    - listen: '127.0.0.1:8082'
  roles.my-role:
    - httpd: additional
```

[socket_ref]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/socket/#socket-tcp-server

## Roles

New roles could be accessed from this project.

> ### `httpd` role

Reviewer: ```suggestion
### `roles.httpd` role
```

This role can be used only with Tarantool 3. It allows configuring one or more
HTTP servers. Those servers could be reused by several other roles.

Example of the configuration:

```yaml
roles_cfg:
  roles.httpd:
    default:
    - listen: 8081
    additional:
    - listen: '127.0.0.1:8082'
```

Server address should be provided either as a URI or as a single port
(in this case, `0.0.0.0` address is used).

User can access every working HTTP server from the configuration by name,

> using `get_server(name)` method.

Reviewer: ```suggestion
using `require('roles.httpd').get_server(name)` method.
```

If the `name` argument is `nil`, the default server is returned

> (its name should be equal to constant `DEFAULT_SERVER_NAME`).

Reviewer: And what is a value of this constant?

Author: `default` (this constant is also accessible from the role variable).

Reviewer: Please, add it to the document. Configuration from the config should ideally be done without code.

Let's look at the example of using this role. Consider a new role `hello_world`:
```lua
local http_client = require('http.client')

local M = {}
local server = {}

M.apply = function(conf, http_server_role)
    server = http_server_role.get_server(conf.httpd).httpd

    server:route({
        path = '/hello/world'
    }, function(tx)
        return tx:render({text = 'Hello, world!'})
    end)
end

M.get = function()
    local server_uri = ('http://%s:%s'):format(server.host, server.port)
    return http_client.get(server_uri .. '/hello/world')

> end

Reviewer: Let's remove the code and just make a request in an example, using `curl`.

return M
```

This role accepts a server by name from a config and creates a route to return 
`Hello, world!` to every request by this route.

Config for `httpd` and `hello_world` might look like this:
```yaml
roles_cfg = {
  httpd = {
    default = {
      listen = 8081,
    },
    additional = {
      listen = '127.0.0.1:8082',
    }
  },
  hello_world = {
    httpd = 'additional',
  },
}
```

After that, we can get `Hello, world!` by just running this script on a
Tarantool instance:
```lua
local httpd_role = require('roles.httpd')
local hello_role = require('roles.hello_world')

httpd_role.apply(config.roles_cfg.httpd)
hello_role.apply(config.roles_cfg.hello_world, httpd_role)

print(hello_role.get().body)

> ```

Reviewer: Nobody starts Tarantool like that, we do it in such way for testing. It's better to replace the example with a real cluster configuration file + `tt start`. See examples in Tarantool's docs: https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets .
